### PR TITLE
Premium Products: allow key=val format for product options

### PR DESCRIPTION
**Pull request by mlutfy**
Overview
----------------------------------------

Adds support for formatting premium/product options as key=value. Makes it easier to manage products where the options might have their own SKU. It also displays the option SKU on ContributionView.

Another small change was to display "Fulfilled : No" when there is no fulfillment date.

Before
----------------------------------------

Product options can have only a label. This is not super useful for inventory management.

For example, if I have t-shirts that have S/M/L, each of those options have their own SKU. In reports, I would see only "S", "M", "L". I cannot put the SKU there, because the string is user-facing (when selecting the option).

After
----------------------------------------

Products can have a key=label, so the user sees "M" but my inventory reports see "SHIRT-2024-M".

Also changes the "Fulfilled" display, to show "No" when not fulfilled, because at a glance, one doesn't know that a date will be displayed here, so it gives the impression that the order was fulfilled:

image

image

Comments
----------------------------------------

I considered moving options to either an OptionValue, or to a separate entity altogether, but my inner voice said it would be very overkill, and I'm pretty sure no one uses this feature, because it's very awkward to use.

There is one drawback, but it is opt-in: in a report of Contributions (or when using SearchKit), we will see the option SKU instead of the label. Fetching the label will not easy in searchkit. However, 1) I think most people will want the SKU, and 2) this behaviour is opt-in, if people previously only had a "label", then it works like before.

These changes came from the Tor Project Association, where they had been using lots of messy custom fields to manage perks/premiums for donations, presumably as a workaround for this limitation.

### CODE REVIEW

File: CRM/Contribute/Form/Contribution.php [hunk 1]
```diff
@@ -512,9 +515,6 @@ public function setDefaultValues() {
       }
     }
 
-    $this->assign('is_pay_later', !empty($defaults['is_pay_later']));
-
-    $this->assign('contribution_status_id', CRM_Utils_Array::value('contribution_status_id', $defaults));
```
Comment (mlutfy) [inline CRM/Contribute/Form/Contribution.php:517]: these lines were moved up, to tidy up a bit

**Comment by mlutfy** [comment]
Test fails:

```
not ok 881 - Error: CRM_Contribute_Form_ContributionTest::testPremiumUpdate
not ok 882 - Error: CRM_Contribute_Form_ContributionTest::testPremiumUpdateCreditCard
```

**Comment by mlutfy** [comment]
Attempting to fix the tests (they run locally, but my fix may be dodgy.. but manual testing indicates it works). Also added a new test for the key=val format.

**Comment by mlutfy** [comment]
Tests pass :tada: 

but let me know if the changes make sense

**Comment by colemanw** [comment]
@mlutfy yes this makes sense to me. I'm happy to merge this, but as a follow-up, it looks like the schema is missing an important bit of metadata. Serialized fields are supposed to declare *how* they are serialized, so that the api can automatically convert them to/from array format.

Here: https://github.com/civicrm/civicrm-core/pull/29767
I don't know if this impacts your PR at all but that will cause the API to return this field as an array instead of as a string.

I think a reasonable followup to that PR would be to add custom input/output formatters to that field in APIv4 to convert this
```
[
  0 => '#123=Shirt',
  1 => '#456=Mug',
]
```
into this:
```
[
  '#123' => 'Shirt',
  '#456' => 'Mug',
]
```

**Comment by mlutfy** [comment]
rebased

**Comment by colemanw** [comment]
@mlutfy sure, here's some:
https://github.com/civicrm/civicrm-core/blob/f695290205a9d0edcd55817cdfaa35096fd0420c/Civi/Api4/Service/Spec/Provider/FileGetSpecProvider.php#L26-L54

Those are all output formatters. For formatting the input, you'd probably do that at the BAO level actually, probably with `hook_civicrm_pre()`.

**Pull request merged**